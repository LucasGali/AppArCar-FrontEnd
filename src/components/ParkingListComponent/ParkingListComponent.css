#parktab {
  text-align: center;
  margin: 0;
}

.search-wrapper {
  display: flex;
  width: 30%;
  font-size: 20px;
  margin: 0 10px 10px;
}

i.search-icon {
  height: calc(1.5em + 0.75rem + 2px);
  padding-left: 6px;
  font-size: 1rem;
  font-weight: 400;
  line-height: 1.5;
  color: #495057;
  background-color: #fff;
  background-clip: padding-box;
  border: 1px solid #ced4da;
  border-radius: 0.25rem;
  transition: border-color 0.15s ease-in-out, box-shadow 0.15s ease-in-out;
  border-right: none;
  border-top-right-radius: 0px;
  border-bottom-right-radius: 0px;
  display: flex;
  align-items: center;
}

i:hover {
  cursor: auto;
}

.search-wrapper input {
  border-left: none;
  border-top-left-radius: 0px;
  border-bottom-left-radius: 0px;
}

.form-control:focus {
  box-shadow: none;
  border-color: #ced4da;
}

.spinner {
  position: relative;
  top: 20vh;
}
