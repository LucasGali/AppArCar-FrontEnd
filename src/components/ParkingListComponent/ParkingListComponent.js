import React, { useMemo, useState, useEffect } from 'react';
import { useTable } from 'react-table';
import ReserveButton from '../Reserve/ReserveButton';
import './ParkingListComponent.css';
import text from 'utils/text';
import iconSet from '../../icons.json';
import IcomoonReact from 'icomoon-react';
import Mapbox from 'components/Mapbox/Mapbox';
import NoParking from 'components/NoParking/NoParking';
import axios from 'axios';
import { useDispatch } from 'react-redux';
import { parkingData } from '../../actions/index';
import { LOAD_PARKINGS } from '../../actions/types';
import { LoopCircleLoading } from 'react-loadingg';

const ParkingListComponent = ({ vehicleData, locationPoint }) => {
  const columnsData = [
    {
      Header: text.parkingTable.columns.name,
      accessor: 'name',
    },
    {
      Header: text.parkingTable.columns.address,
      accessor: 'address',
    },
    {
      Header: text.parkingTable.columns.attention,
      accessor: 'business_hours',
    },
    {
      Header: text.parkingTable.columns.price,
      accessor: 'price',
    },
  ];
  // eslint-disable-next-line react-hooks/exhaustive-deps
  const columns = useMemo(() => columnsData, []);
  const [search, setSearch] = useState('');
  const [data, setData] = useState([]);
  const [dataToShow, setDataToShow] = useState([]);
  const [spinner, setSpinner] = useState(true);
  const dispatch = useDispatch();

  const tableInstance = useTable({
    columns,
    data,
  });

  const getData = async () => {
    try {     
      const headers = {
        Authorization: localStorage.jwt,
      };
      const response = await axios.get(
        `${process.env.REACT_APP_SERVER_URL}/parkings`,        
        { headers },
      );
      const toShow = response.data
        ?.filter((lot) => {
          const cleaner = (coordinates) => Number(coordinates?.toString()?.slice(0, 10));

          const latitude = locationPoint.coords?.latitude;
          const longitude = locationPoint.coords?.longitude;
          // 0.001Â° = 111 m mts to long & lat
          const meters = 0.01;
          if (
            lot.latitude < cleaner(latitude) + meters &&
            lot.latitude > cleaner(latitude) - meters &&
            lot.longitude < cleaner(longitude) + meters &&
            lot.longitude > cleaner(longitude) - meters
          )
            return lot;
          return undefined
        })
        ?.map(lot => {
          return {
              _id: lot._id,
              name: lot.name,
              address: lot.address,
              business_hours: lot.business_hours,
              price: lot.price[vehicleData.vehicleType],
            };
        });
      dispatch(parkingData(LOAD_PARKINGS, response.data));
      setData([...toShow]);
      setDataToShow([...toShow]);
      setTimeout(() => setSpinner(false), 1000);
    } catch (e) {
      console.error('Oopps!', e);
    }
  };

  useEffect(() => {
    getData();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [vehicleData]);

  const handleChange = (input) => {
    const filterItems = (element) => {
      const parkingName = element.name;
      const parkingAddress = element.address;
      const predicate = input.target.value?.toLowerCase();
      return (
        parkingName?.toLowerCase().includes(predicate) ||
        parkingAddress?.toLowerCase().includes(predicate)
      );
    };

    setSearch(input.target.value);
    if (input.target.value.length > 0) {
      let helper = dataToShow.filter(filterItems);
      setData([...helper]);
    } else {
      setData([...dataToShow]);
    }
  };

  const SearchIcon = () => (
    <IcomoonReact iconSet={iconSet} color='#444' size={20} icon='search' />
  );

  const { getTableProps, getTableBodyProps, headerGroups, rows, prepareRow } =
    tableInstance;
  
  const Table = () => {
    return (<table
      id='parktab'
      className='table table-striped table-dark'
      {...getTableProps()}
    >
      <thead>
        {headerGroups.map((headerGroup) => (
          <tr {...headerGroup.getHeaderGroupProps()}>
            {headerGroup.headers.map((column) => (
              <th {...column.getHeaderProps()}>{column.render('Header')}</th>
            ))}
          </tr>
        ))}
      </thead>
      <tbody {...getTableBodyProps()}>
        {rows.map((row) => {
          prepareRow(row);
          return (
            <tr key={Math.random()} {...row.getRowProps()}>
              {row.cells.map((cell) => (
                <td key={Math.random()} {...cell.getCellProps()}>
                  {cell.render('Cell')}
                </td>
              ))}
              <td>
                <ReserveButton selectedParking={row.original._id} />
              </td>
            </tr>
          );
        })}
      </tbody>
    </table>
    );
  };

  return (
    <>
      <div className='search-wrapper'>
        <i className='search-icon'>
          <SearchIcon />
        </i>
        <input
          type='text'
          className='form-control'
          placeholder={text.home.table.searchInput}
          value={search}
          onChange={handleChange}
        />
      </div>
      <div className='home-map-wrapper'>
        <Mapbox />
      </div>
      {spinner ? (
        <div className='spinner'>
          <LoopCircleLoading />
        </div>
      ) : data.length === 0 ? (
        <NoParking />
      ) : (
        <Table />
      )}
    </>
  );
};

export default ParkingListComponent;
