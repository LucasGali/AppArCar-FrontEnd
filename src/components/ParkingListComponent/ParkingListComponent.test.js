import { render, screen, fireEvent } from '@testing-library/react';
import ParkingListComponent from "./ParkingListComponent"

describe("Parking List Component", () => {
  it("Renders table according to snapshot", () => {
    const { asFragment } = render(<ParkingListComponent />);
    expect(asFragment(<ParkingListComponent />)).toMatchSnapshot();
  })
})
