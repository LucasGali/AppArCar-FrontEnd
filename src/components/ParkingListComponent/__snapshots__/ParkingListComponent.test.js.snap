// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Parking List Component Renders table according to snapshot 1`] = `
<DocumentFragment>
  <div
    class="search-wrapper"
  >
    <i
      class="search-icon"
    >
      <svg
        class=""
        height="20"
        style="display: inline-block; vertical-align: middle;"
        viewBox="0 0 1024 1024"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M992.262 871.396l-242.552-206.294c-25.074-22.566-51.89-32.926-73.552-31.926 57.256-67.068 91.842-154.078 91.842-249.176 0-212.078-171.922-384-384-384-212.076 0-384 171.922-384 384s171.922 384 384 384c95.098 0 182.108-34.586 249.176-91.844-1 21.662 9.36 48.478 31.926 73.552l206.294 242.552c35.322 39.246 93.022 42.554 128.22 7.356s31.892-92.898-7.354-128.22zM384 640c-141.384 0-256-114.616-256-256s114.616-256 256-256 256 114.616 256 256-114.614 256-256 256z"
          style="fill: #444;"
        />
      </svg>
    </i>
    <input
      class="form-control"
      placeholder="Buscar por nombre y/o direccion"
      type="text"
      value=""
    />
  </div>
  <table
    class="table table-striped table-dark"
    id="parktab"
    role="table"
  >
    <thead>
      <tr
        role="row"
      >
        <th
          colspan="1"
          role="columnheader"
        >
          Nombre
        </th>
        <th
          colspan="1"
          role="columnheader"
        >
          Direccion
        </th>
        <th
          colspan="1"
          role="columnheader"
        >
          Horario de Atencion
        </th>
        <th
          colspan="1"
          role="columnheader"
        >
          Precio
        </th>
        <th
          colspan="1"
          role="columnheader"
        >
          Favorito
        </th>
      </tr>
    </thead>
    <tbody
      role="rowgroup"
    >
      <tr
        role="row"
      >
        <td
          role="cell"
        >
          Parking Tucuman
        </td>
        <td
          role="cell"
        >
          25 de Mayo 312
        </td>
        <td
          role="cell"
        >
          24 hs
        </td>
        <td
          role="cell"
        />
        <td
          role="cell"
        >
          25
        </td>
        <td>
          <div>
            <button
              class="btn-openModal"
            >
              Reserva
            </button>
          </div>
        </td>
      </tr>
      <tr
        role="row"
      >
        <td
          role="cell"
        >
          La Teresina
        </td>
        <td
          role="cell"
        >
          Mendoza 949
        </td>
        <td
          role="cell"
        >
          8 - 21 hs
        </td>
        <td
          role="cell"
        />
        <td
          role="cell"
        >
          30
        </td>
        <td>
          <div>
            <button
              class="btn-openModal"
            >
              Reserva
            </button>
          </div>
        </td>
      </tr>
      <tr
        role="row"
      >
        <td
          role="cell"
        >
          Estacionamiento Playa Grande
        </td>
        <td
          role="cell"
        >
          Provincia de Cordoba 751
        </td>
        <td
          role="cell"
        >
          7 - 22 hs
        </td>
        <td
          role="cell"
        />
        <td
          role="cell"
        >
          35
        </td>
        <td>
          <div>
            <button
              class="btn-openModal"
            >
              Reserva
            </button>
          </div>
        </td>
      </tr>
      <tr
        role="row"
      >
        <td
          role="cell"
        >
          Estacionamiento del Mercado
        </td>
        <td
          role="cell"
        >
          Maipú 246
        </td>
        <td
          role="cell"
        >
          24 hs
        </td>
        <td
          role="cell"
        />
        <td
          role="cell"
        >
          40
        </td>
        <td>
          <div>
            <button
              class="btn-openModal"
            >
              Reserva
            </button>
          </div>
        </td>
      </tr>
      <tr
        role="row"
      >
        <td
          role="cell"
        >
          Estacionamiento SM
        </td>
        <td
          role="cell"
        >
          Gral. José de San Martín 867
        </td>
        <td
          role="cell"
        >
          8 - 22 hs
        </td>
        <td
          role="cell"
        />
        <td
          role="cell"
        >
          35
        </td>
        <td>
          <div>
            <button
              class="btn-openModal"
            >
              Reserva
            </button>
          </div>
        </td>
      </tr>
      <tr
        role="row"
      >
        <td
          role="cell"
        >
          Estacionamiento 970
        </td>
        <td
          role="cell"
        >
          Crisóstomo Álvarez 970 
        </td>
        <td
          role="cell"
        >
          8 - 20 hs
        </td>
        <td
          role="cell"
        />
        <td
          role="cell"
        >
          25
        </td>
        <td>
          <div>
            <button
              class="btn-openModal"
            >
              Reserva
            </button>
          </div>
        </td>
      </tr>
      <tr
        role="row"
      >
        <td
          role="cell"
        >
          Guardeía 9 de Julio
        </td>
        <td
          role="cell"
        >
          Pcia. de Córdoba 950 
        </td>
        <td
          role="cell"
        >
          24 hs
        </td>
        <td
          role="cell"
        />
        <td
          role="cell"
        >
          35
        </td>
        <td>
          <div>
            <button
              class="btn-openModal"
            >
              Reserva
            </button>
          </div>
        </td>
      </tr>
      <tr
        role="row"
      >
        <td
          role="cell"
        >
          Estacionamiento E
        </td>
        <td
          role="cell"
        >
          25 de Mayo 201
        </td>
        <td
          role="cell"
        >
          7 - 22 hs
        </td>
        <td
          role="cell"
        />
        <td
          role="cell"
        >
          30
        </td>
        <td>
          <div>
            <button
              class="btn-openModal"
            >
              Reserva
            </button>
          </div>
        </td>
      </tr>
      <tr
        role="row"
      >
        <td
          role="cell"
        >
          Estacionamiento Ayacucho
        </td>
        <td
          role="cell"
        >
          Batalla de Ayacucho 53
        </td>
        <td
          role="cell"
        >
          8 - 22 hs
        </td>
        <td
          role="cell"
        />
        <td
          role="cell"
        >
          35
        </td>
        <td>
          <div>
            <button
              class="btn-openModal"
            >
              Reserva
            </button>
          </div>
        </td>
      </tr>
      <tr
        role="row"
      >
        <td
          role="cell"
        >
          Estacionamiento Playa 24
        </td>
        <td
          role="cell"
        >
          Av. 24 de Septiembre 771
        </td>
        <td
          role="cell"
        >
          24 hs
        </td>
        <td
          role="cell"
        />
        <td
          role="cell"
        >
          40
        </td>
        <td>
          <div>
            <button
              class="btn-openModal"
            >
              Reserva
            </button>
          </div>
        </td>
      </tr>
      <tr
        role="row"
      >
        <td
          role="cell"
        >
          Parking Tucuman
        </td>
        <td
          role="cell"
        >
          25 de Mayo 312
        </td>
        <td
          role="cell"
        >
          24 hs
        </td>
        <td
          role="cell"
        />
        <td
          role="cell"
        >
          25
        </td>
        <td>
          <div>
            <button
              class="btn-openModal"
            >
              Reserva
            </button>
          </div>
        </td>
      </tr>
      <tr
        role="row"
      >
        <td
          role="cell"
        >
          La Teresina
        </td>
        <td
          role="cell"
        >
          Mendoza 949
        </td>
        <td
          role="cell"
        >
          8 - 21 hs
        </td>
        <td
          role="cell"
        />
        <td
          role="cell"
        >
          30
        </td>
        <td>
          <div>
            <button
              class="btn-openModal"
            >
              Reserva
            </button>
          </div>
        </td>
      </tr>
      <tr
        role="row"
      >
        <td
          role="cell"
        >
          Estacionamiento Playa Grande
        </td>
        <td
          role="cell"
        >
          Provincia de Cordoba 751
        </td>
        <td
          role="cell"
        >
          7 - 22 hs
        </td>
        <td
          role="cell"
        />
        <td
          role="cell"
        >
          35
        </td>
        <td>
          <div>
            <button
              class="btn-openModal"
            >
              Reserva
            </button>
          </div>
        </td>
      </tr>
      <tr
        role="row"
      >
        <td
          role="cell"
        >
          Estacionamiento del Mercado
        </td>
        <td
          role="cell"
        >
          Maipú 246
        </td>
        <td
          role="cell"
        >
          24 hs
        </td>
        <td
          role="cell"
        />
        <td
          role="cell"
        >
          40
        </td>
        <td>
          <div>
            <button
              class="btn-openModal"
            >
              Reserva
            </button>
          </div>
        </td>
      </tr>
      <tr
        role="row"
      >
        <td
          role="cell"
        >
          Estacionamiento SM
        </td>
        <td
          role="cell"
        >
          Gral. José de San Martín 867
        </td>
        <td
          role="cell"
        >
          8 - 22 hs
        </td>
        <td
          role="cell"
        />
        <td
          role="cell"
        >
          35
        </td>
        <td>
          <div>
            <button
              class="btn-openModal"
            >
              Reserva
            </button>
          </div>
        </td>
      </tr>
      <tr
        role="row"
      >
        <td
          role="cell"
        >
          Estacionamiento 970
        </td>
        <td
          role="cell"
        >
          Crisóstomo Álvarez 970 
        </td>
        <td
          role="cell"
        >
          8 - 20 hs
        </td>
        <td
          role="cell"
        />
        <td
          role="cell"
        >
          25
        </td>
        <td>
          <div>
            <button
              class="btn-openModal"
            >
              Reserva
            </button>
          </div>
        </td>
      </tr>
      <tr
        role="row"
      >
        <td
          role="cell"
        >
          Guardeía 9 de Julio
        </td>
        <td
          role="cell"
        >
          Pcia. de Córdoba 950 
        </td>
        <td
          role="cell"
        >
          24 hs
        </td>
        <td
          role="cell"
        />
        <td
          role="cell"
        >
          35
        </td>
        <td>
          <div>
            <button
              class="btn-openModal"
            >
              Reserva
            </button>
          </div>
        </td>
      </tr>
      <tr
        role="row"
      >
        <td
          role="cell"
        >
          Estacionamiento E
        </td>
        <td
          role="cell"
        >
          25 de Mayo 201
        </td>
        <td
          role="cell"
        >
          7 - 22 hs
        </td>
        <td
          role="cell"
        />
        <td
          role="cell"
        >
          30
        </td>
        <td>
          <div>
            <button
              class="btn-openModal"
            >
              Reserva
            </button>
          </div>
        </td>
      </tr>
      <tr
        role="row"
      >
        <td
          role="cell"
        >
          Estacionamiento Ayacucho
        </td>
        <td
          role="cell"
        >
          Batalla de Ayacucho 53
        </td>
        <td
          role="cell"
        >
          8 - 22 hs
        </td>
        <td
          role="cell"
        />
        <td
          role="cell"
        >
          35
        </td>
        <td>
          <div>
            <button
              class="btn-openModal"
            >
              Reserva
            </button>
          </div>
        </td>
      </tr>
      <tr
        role="row"
      >
        <td
          role="cell"
        >
          Estacionamiento Playa 24
        </td>
        <td
          role="cell"
        >
          Av. 24 de Septiembre 771
        </td>
        <td
          role="cell"
        >
          24 hs
        </td>
        <td
          role="cell"
        />
        <td
          role="cell"
        >
          40
        </td>
        <td>
          <div>
            <button
              class="btn-openModal"
            >
              Reserva
            </button>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
